test: add new python linter to check file names and permissions

windsok opened the pull request:
Adds a new python linter test which tests for correct filenames and file permissions in the repository.

Replaces the existing tests in the `test/lint/lint-filenames.sh` and `test/lint/lint-shebang.sh` linter tests, as well as adding some new and increased testing. This increased coverage is intended to catch issues such as in #21728 and https://github.com/bitcoin/bitcoin/pull/16807/files#r345547050

Summary of tests:
* Checks every file in the repository against an allowed regexp to make sure only lowercase or uppercase alphanumerics (a-zA-Z0-9), underscores (_), hyphens (-), at (@) and dots (.) are used in repository filenames.

* Checks only source files (*.cpp, *.h, *.py, *.sh) against a stricter allowed regexp to make sure only lowercase alphanumerics (a-z0-9), underscores (_), hyphens (-) and dots (.) are used in source code filenames. Additionally there is an exception regexp for directories or files which are excepted from matching this regexp (This should replicate the existing `test/lint/lint-filenames.sh` test)

* Checks all files in the repository match an allowed executable or non-executable file permission octal. Additionally checks that for executable files, the file contains a shebang line.

* Checks that for executable `.py` and `.sh` files, the shebang line used matches an allowable list of shebangs (This should replicate the existing `test/lint/lint-shebang.sh` test)

* Checks every file that contains a shebang line to ensure it has an executable permission

Additionally updates the permissions on various files to comply with the new tests.

Fixes #21729

windsok committed:
test: add new python linter to check file names and permissions

Replaces the existing tests in the test/lint/lint-filenames.sh and test/lint/lint-shebang.sh linter tests, as well as adding some new and increased testing.
Summary of tests:
- Checks every file in the repository against an allowed regexp to make sure only lowercase or uppercase alphanumerics (a-zA-Z0-9), underscores (_), hyphens (-), at (@) and dots (.) are used in repository filenames.
- Checks only source files (*.cpp, *.h, *.py, *.sh) against a stricter allowed regexp to make sure only lowercase alphanumerics (a-z0-9), underscores (_), hyphens (-) and dots (.) are used in source code filenames. Additionally there is an exception regexp for directories or files which are excepted from matching this regexp (This should replicate the existing test/lint/lint-filenames.sh test)
- Checks all files in the repository match an allowed executable or non-executable file permission octal. Additionally checks that for executable files, the file contains a shebang line.
- Checks that for executable .py and .sh files, the shebang line used matches an allowable list of shebangs (This should replicate the existing test/lint/lint-shebang.sh test)
- Checks every file that contains a shebang line to ensure it has an executable permission

Fixes #21729

practicalswift commented on test/lint/lint-filenames-permissions.py:
```diff
@@ -0,0 +1,141 @@
+
+CMD_ALL_FILES = "git ls-files --full-name"
+CMD_SOURCE_FILES = 'git ls-files --full-name -- "*.[cC][pP][pP]" "*.[hH]" "*.[pP][yY]" "*.[sS][hH]"'
+CMD_SHEBANG_FILES = """git grep --full-name --line-number -I '^#!' | grep :1: | cut -d ":" -f 1"""
```
Nit: Consider doing the filtering (`grep`, `cut`) in Python instead to make the script easier to read and maintain. In the Python code a comment could be added that `:1:` is used to make sure only shebangs on the first line are considered.

windsok replied:
updated to move the filtering into python :)

practicalswift commented on test/lint/lint-filenames-permissions.py:
```diff
@@ -0,0 +1,141 @@
+    files = check_output(CMD_ALL_FILES, shell=True).decode("utf8").strip().split("\n")
+    filename_regex = re.compile(ALLOWED_FILENAME_REGEXP)
+    exit_code = 0
+    for file in files:
```
Nit: `file` is a Python builtin. Consider using `filename/filenames` instead of `file/files` :)

windsok replied:
good call, updated

practicalswift commented on test/lint/lint-filenames-permissions.py:
```diff
@@ -0,0 +1,141 @@
+        file_perms = int(oct(os.stat(file).st_mode)[-3:])
+        if file_perms != ALLOWED_PERMISSION_EXECUTABLES:
+            print(
+                f"""File "{file}" contains a shebang line, but has the file permission {file_perms} instead of the expected executable permission {ALLOWED_PERMISSION_EXECUTABLES}. Do "chmod {ALLOWED_PERMISSION_EXECUTABLES} {file}" (or remove the shebang line)."""
```
This requirement holds for shebang files that expected to be executed directly.

To avoid this requirement for files that are not expected to be excluded directly but still have shebangs consider adding exclusions for:
* `*.py` files that do not contain `if __name__ == '__main__':` or `if __name__ == "__main__":`: these are typically modules that are not expected to be executed directly.
* `*.bash` files: these are typically expected to be sourced and not executed directly.

It could be argued that shebangs should be removed from these types of files, but I think shebangs could still be useful to allow for language/dialect auto-detection in editors, `shellcheck`, etc.

windsok replied:
Updated. I wonder if some other file types should also be skipped along with `*bash`, I notice the test is failing for some files like `contrib/init/bitcoind.init`, `contrib/init/bitcoind.openrc`, `contrib/devtools/split-debug.sh.in"` and I'm thinking they might fit into the same category of files which are sourced but not executed?

There is also quite a few `.sh` files in `ci/test/` like `ci/test/00_setup_env_android.sh` which fail the test, but I'm not sure if those really should be forced to be executable or not?

Thanks very much for the initial review and feedback :)

practicalswift replied:
> Updated. I wonder if some other file types should also be skipped along with `*bash`? I notice the test is failing for some files like `contrib/init/bitcoind.init`, `contrib/init/bitcoind.openrc`, `contrib/devtools/split-debug.sh.in` and I'm thinking they might fit into the same category of files which are sourced but not executed?

Sounds reasonable. Skipping the requirement for `*.{init,openrc,sh.in}` should be fine.

> There is also quite a few `.sh` files in `ci/test/` like `ci/test/00_setup_env_android.sh` which fail the test, but I'm not sure if those really should be forced to be executable or not?

Perhaps we should simply make these executable to avoid the complexity of making the `+x` requirement depend also on directory (in addition to `#!` presence and suffix).

windsok replied:
Thanks! updated the condition to skip the requirement for `*.{init,openrc,sh.in}` files.

For the remaining [files failing the test](https://github.com/bitcoin/bitcoin/pull/21740/checks), what would be the correct way to go about updating those to comply with the test? Should it be a second commit on this PR?

Thanks again for the review and guidance.

practicalswift reviewed (changes requested):
Concept ACK: desired properties should where possible be checked automatically.

Linters are to "repo invariants" what constructors are to class invariants :)

Nice to see the linting being done in Python instead of bash: much better in terms of maintainability, portability and readability.

kiminuo commented on test/lint/lint-filenames-permissions.py:
```diff
@@ -0,0 +1,160 @@
+
+    Additionally there is an exception regexp for directories or files which are excepted from matching this regexp.
+    """
+    filenames = check_output(CMD_SOURCE_FILES, shell=True).decode("utf8").strip().split("\n")
```
Out of curiosity: Would it make sense to replace `strip()` with `rstrip()` (https://docs.python.org/3/library/stdtypes.html?highlight=rstrip#str.rstrip)?

Just for fun: There is still one very specific case where this does not work (I hope I'm not missing something): You can add a file like this `git add test/util/zzz.py  ` (note the trailing whitespace) and `strip()` will make sure that `test/util/zzz.py` will be processed and not `test/util/zzz.py  `. I think your code is fine but if it can be improved wrt robustness I'm all in (possibly split first and remove trailing empty entries?)

windsok replied:
Great catch!

I've updated the logic to not use a `strip()` at all, and instead split on the newlines and then remove the trailing empty element. I chose to do this using a list comprehension filtering for empty list elements, rather than a `.pop()` of the last element, as it seemed slightly more robust, but can change it to a `pop()` if I'm overthinking it.

I tested that this does now catch files with trailing whitespace:
```bash
File "test/util/rpcauth-test.py " does not not match the allowed filename regexp ('^[a-zA-Z0-9/_.@][a-zA-Z0-9/_.@-]*$').
```

kiminuo commented on test/lint/lint-filenames-permissions.py:
```diff
@@ -0,0 +1,160 @@
+        file_perms = int(oct(os.stat(filename).st_mode)[-3:])
+
+        if file_perms == ALLOWED_PERMISSION_EXECUTABLES:
+            shebang = open(filename, "rb").readline().strip()
```
I believe the `strip()` here is not correct. I don't think that you can put whitespace before shebang. 

It's a question whether one should allow whitespace after shebank AFAIK it's allowed but ... It may be for the best simply remove `.strip()`.

Just for fun: https://unix.stackexchange.com/questions/276751/is-space-allowed-between-and-bin-bash-in-shebang

windsok replied:
Great catch, thanks.

I updated this to use `rstrip(b"\n")` to remove only the trailing newline

windsok commented on test/lint/lint-filenames-permissions.py:
```diff
@@ -0,0 +1,160 @@
+
+CMD_ALL_FILES = "git ls-files --full-name"
+CMD_SOURCE_FILES = 'git ls-files --full-name -- "*.[cC][pP][pP]" "*.[hH]" "*.[pP][yY]" "*.[sS][hH]"'
+CMD_SHEBANG_FILES = """git grep --full-name --line-number -I '^#!'"""
```
Thanks, updated :)

fanquake commented:
Can you update the commit message to use the following format:
```bash
prefix: title

commit body
```

i.e:
```bash
test: add new python linter to check file names and permissions

explain what is being done in this commit
explain some more etc
```

windsok committed:
test: fix file permissions on various scripts

Updates permissions on files to comply with the new test added in the following commit

windsok commented:
> Can you update the commit message to use the following format:
> 
Thanks @fanquake  - I've updated the commit messages to the standard format

> Also, for this to me merged. Obviously all [linters](https://github.com/bitcoin/bitcoin/pull/21740/checks?check_run_id=2415315623) / CI need to be passing. So if changes to other files need to be made, they will have to be done either in preceding commits, or as part of the same change.

Done - added a new commit before the main commit which updates the permissions on files which are failing the new test.

Would appreciate any additional review and feedback :)

practicalswift commented:
cr ACK 52c3fb3c9d4145819bd0fe839ba89482e73fce99: patch looks correct!

Nice first-time contribution @windsok. I hope to see more contributions from you in the future. Warm welcome as a contributor!

kiminuo commented:
@windsok This PR removes `test/lint/lint-shebang.sh` and `test/lint/lint-filenames.sh` and adds `test/lint/lint-filenames-permissions.sh` and adds `test/lint/lint-filenames-permissions.py`. 

The new file names are confusing to me as they do something different than they hint. Would it make sense to you to rename them from "lint-filenames-permissions.{py|sh}" to "lint-files.{py|sh}" or something better?

kiminuo commented on test/lint/lint-filenames-permissions.py:
```diff
@@ -0,0 +1,182 @@
+    failed_tests += check_shebang_file_permissions()
+
+    if failed_tests:
+        sys.exit(1)
```
Nit: Given that `failed_tests` number is computed, one can probably print the number of failed tests too.

windsok replied:
Great idea, I've updated to keep track of the number of failures, and print a message with the total number of failures, if any :)

kiminuo commented on test/lint/lint-filenames-permissions.py:
```diff
@@ -0,0 +1,182 @@
+    'ci/retry/retry' -> None
+    'contrib/devtools/split-debug.sh.in' -> 'sh.in'
+    """
+    filename_parts = filename.split(os.extsep, 1)
```
I think people mostly do something like this https://source.dot.net/#System.Private.CoreLib/Path.cs,467ebc0e33e0820c to find a file extension. That is to look for the first `os.extsep` but not from the beginning of the string (filename) but from the end of the filename.

A counter-example can be `'contrib/devtools/split-debug.my.nice.helper.sh.in'` where you would return `my.nice.helper.sh.in`.

Your code may be perfectly fine for what you want to achieve in this PR, yet I would be hesitant to use `get_filename_extension` name. 

I have no clear suggestion here, letting it up to you to decide. :)

windsok replied:
Ah yes I originally was using `os.path.splitext(self.file_path)[1].strip(".")` which works similar to the linked implementation (only return the final extension, not everything from the first `.`)

I had changed this as I needed a way to detect `.sh.in` files to exclude them from one of the tests, but you bring up a very good point - with the change I had made it's possible certain files with weird extensions like `split-debug.my.nice.helper.sh` would not actually be detected as a `sh` file, and would not have certain tests enforced.

Ive updated the code to add a new `FileMeta` class, and have changed the logic to use the appropriate version of the extension in each test. Let me know if this makes sense or not :)

kiminuo replied:
It looks fine to me. Thanks for updating it. :)

windsok commented:
> @windsok This PR removes `test/lint/lint-shebang.sh` and `test/lint/lint-filenames.sh` and adds `test/lint/lint-filenames-permissions.sh` and adds `test/lint/lint-filenames-permissions.py`.
> 
> The new file names are confusing to me as they do something different than they hint. Would it make sense to you to rename them from "lint-filenames-permissions.{py|sh}" to "lint-files.{py|sh}" or something better?

I'm happy to name this test as `lint-files.{py|sh}` - it does seem clearer and less confusing. Thanks for the suggestion :)

> * This PR modifies `https://github.com/bitcoin/bitcoin/blob/master/contrib/gitian-descriptors/assign_DISTNAME` permissions but it may be actually correct to add `.sh` extension - I'm not sure. Just mentioning it here so that others double check correctness of the change.

I am also unsure on this one, but happy to take advice on which direction it should go

windsok commented:
> Nice first-time contribution @windsok. I hope to see more contributions from you in the future. Warm welcome as a contributor!

Thanks @practicalswift :) looking forward to getting more involved. Sorry for the delayed response on the last round of feedback.

practicalswift commented:
cr re-ACK 46b025e00df40724175735eb5606ac73067cb3b8: patch still looks correct

kiminuo commented on test/lint/lint-files.py:
```diff
@@ -0,0 +1,207 @@
+    """
+    # We avoid using rstrip() to ensure we catch filenames which accidentally include trailing whitespace
+    filenames = check_output(CMD_ALL_FILES, shell=True).decode("utf8").split("\n")
+    filenames = [filename for filename in filenames if filename != ""]  # removes the trailing empty list element
```
nit: The comment does not correspond with the code actually. 

One could probably do: `filenames = filenames[:-1]`

windsok replied:
right, I chose to remove the trailing element by checking for an empty string as I was worried about some case in the future where the `git ls-files` command changes somehow and the last element is not actually empty. I'm probably overthinking it, though.

windsok replied:
On a review of this, I'm fairly certain I can just safely rewrite this code as this one liner, which still catches the original edge case of a file with whitespace at the end of the files list.

```python
filenames = check_output(CMD_ALL_FILES, shell=True).decode("utf8").rstrip('\n').split("\n")
```

Happy to update this if people don't mind re-reviewing. Let me know.

kiminuo replied:
> right, I chose to remove the trailing element by checking for an empty string as I was worried about some case in the future where the output of the `git ls-files` command changes somehow and the last element is not actually empty. I'm probably overthinking it though.

One would need to ask on a Git mailing list whether the text output of the command is stable or not. But your solution works well, so it's good as it is I think. 👍

MarcoFalke replied:
If a filename had a newline in its name it would also be split by the `split("\n")`

windsok replied:
I'll think about a way to handle this case. I think currently if this happens the test overall will still fail, as the permissions test would not be able to locate the file and should error out - we can definitely handle this better, though.

MarcoFalke replied:
I think `.rstrip('\n').split("\n")` as you suggested is fine

windsok replied:
Addressed in https://github.com/bitcoin/bitcoin/pull/21873

kiminuo commented on test/lint/lint-files.py:
```diff
@@ -0,0 +1,207 @@
+        """
+        Returns the octal file permission of the file
+        """
+        return int(oct(os.stat(self.file_path).st_mode)[-3:])
```
super nit: I'm not sure whether `os.stat` should behave have [follow_symlinks](https://docs.python.org/3/library/os.html?highlight=os%20stat#os.stat) equal to `True` (default) or `False`. Possibly relevant is https://stackoverflow.com/questions/954560/how-does-git-handle-symbolic-links/18791647#18791647.

I don't want to hold up this PR. Leaving this for comment here for the safe of completeness.

windsok replied:
Hmm interesting. It seems that there are currently no symbolic links in the repository itself, at least none that I could find with:
```bash
ls -lR ~/code/bitcoin | grep ^l
```

I wonder if that is intentional, or it's just never come up before?

In any case, if there were symbolic links my thinking is that we probably would want to follow them, as it seems that symlinks themselves [can't actually have their permissions changed](https://www.cyberciti.biz/faq/can-i-set-the-permissions-on-a-symbolic-link-under-linuxunix/), so either we would want to exclude symlinks entirely from the permissions test, or just do the default here and follow the link to check the linked file.

kiminuo replied:
My take on this is: I think that adding a symbolic link should produce a warning so that others can notice the fact. The fact that there are no symbolic links at the moment hints that they are probably not necessary.

However, this is really out of scope of this PR, I guess. If somebody finds it important it's always possible to address this in a follow up PR. :)

kiminuo commented:
code review ACK 46b025e00df40724175735eb5606ac73067cb3b8 if `contrib/gitian-descriptors/assign_DISTNAME` permission change is deemed OK.

Great contribution!

laanwj commented:
Code review ACK 46b025e00df40724175735eb5606ac73067cb3b8
Also ACK on the permission changes in 46b025e00df40724175735eb5606ac73067cb3b8.
Thanks for working on this.

> Nice to see the linting being done in Python instead of bash: much better in terms of maintainability, portability and readability.

Agree. Shell script tends to become virtually unreviewable quickly when the logic gets more complex than simply executing some commands in a batch.

laanwj commented on test/lint/lint-files.sh:
```diff
@@ -0,0 +1,7 @@
+
+set -e
+cd "$(dirname $0)/../.."
+test/lint/lint-files.py
```
Just a question (not a blocker): why does this need a bash wrapper?

windsok replied:
I believe the main reason is the the `lint-all.sh` script that CI runs is currently setup to automatically run all `.sh` tests in the `test/lint` directory, but won't automatically run `.py` files.

In addition it allows lint tests [to opt in or out of locale dependence](https://github.com/bitcoin/bitcoin/blob/7fcf53f7b4524572d1d0c9a5fdc388e87eb02416/test/lint/lint-all.sh#L10-L12)

If we start to move more towards using python based linters, it probably makes sense to make some updates here, so we can get rid of the  requirement for the `.sh` wrapper.

@practicalswift may have more context here.

MarcoFalke commented on test/lint/lint-files.py:
```diff
@@ -0,0 +1,207 @@
+    for filename in filenames:
+        file_meta = FileMeta(filename)
+        if file_meta.permissions == ALLOWED_PERMISSION_EXECUTABLES:
+            shebang = open(filename, "rb").readline().rstrip(b"\n")
```
nit: shouldn't `open` be used in a context manager, so that it is closed again?

windsok replied:
Good call, I will address in a follow-up

MarcoFalke commented on test/lint/lint-files.py:
```diff
@@ -0,0 +1,207 @@
+            shebang = open(filename, "rb").readline().rstrip(b"\n")
+
+            # For any file with executable permissions the first line must contain a shebang
+            if shebang[:2] != b"#!":
```
nit: Doesn't matter here, but it could use `startswith` for type safety.

ryanofsky commented on test/lint/lint-files.py:
```diff
@@ -0,0 +1,207 @@
+ALLOWED_SOURCE_FILENAME_EXCEPTION_REGEXP = (
+    "^src/(secp256k1/|univalue/|test/fuzz/FuzzedDataProvider.h)"
+)
+ALLOWED_PERMISSION_NON_EXECUTABLES = 644
+ALLOWED_PERMISSION_EXECUTABLES = 755
```
Since this PR, when I run the linter I see a flood of errors and have to disable it. It seems inappropriate to check user, group, read, and write permissions when the only permission git tracks is whether the file is marked executable. Other permissions are local to the checkout and don't affect the project or other developers. Errors I see are:

```
[...]
File "test/lint/lint-whitespace.sh" contains a shebang line, but has the file permission 700 instead of the expected executable permission 755. Do "chmod 755 test/lint/lint-whitespace.sh" (or remove the shebang line).
File "test/util/bitcoin-util-test.py" contains a shebang line, but has the file permission 700 instead of the expected executable permission 755. Do "chmod 755 test/util/bitcoin-util-test.py" (or remove the shebang line).
File "test/util/rpcauth-test.py" contains a shebang line, but has the file permission 700 instead of the expected executable permission 755. Do "chmod 755 test/util/rpcauth-test.py" (or remove the shebang line).
ERROR: There were 2464 failed tests in the lint-files.py lint test. Please resolve the above errors.
^---- failure generated from test/lint/lint-files.sh
```
